### PR TITLE
Poll organization events

### PR DESCRIPTION
**Pull request by g-arjones**
Here we go... 

This has most of the functionality we've discussed. If I am being completely honest, I am not very happy with our design choice regarding event polling (specially after finding out about https://github.com/rock-core/autoproj-daemon/commit/1910da5c57f965462cbe4556bf9cce86d627cd0b). It seems to me that the efficiency improvement is not worth the HUGE increase in complexity due to several corner cases regarding event ordering and so forth now that we are caching results. Personally, I would gladly just throw away all the event polling thing and do branch synchronization periodically (and implement polling mainline heads to detect pushes to mainline). Besides the obvious benefit of reducing A LOT of code, we also won't have to implement workarounds to deal with the things that we can't get from events (base branch changes in PRs and PR body updates, so far).

That being said, I believe this is good enough to do some testing "in the wild" (even though I would appreciate if you could "stress-test" it before, trying to catch some more corner cases I didn't think of).

I am opening another PR for the overrides generation just because I pretty much did that completely separate from event polling (it may also make easier for review).

### CODE REVIEW

**Comment by doudou** [comment]
The problem with polling repositories directly is that we would have to be very slow to avoid ending up banned by GitHub. Polling repositories simply does not scale, listening to events does - even more since "no events" does not count against rate limiting - i.e. we can poll pretty fast until something happens.

**Comment by g-arjones** [comment]
We are using eTags for everything so nothing counts against the rate limit unless something changes.. Take a look on faraday-http-cache gem for more on that

File: lib/autoproj/cli/daemon.rb [hunk 1]
```diff
@@ -37,76 +49,189 @@ def resolve_packages
...
             # Whether an attempt to update the workspace has failed
             #
             # @return [Boolean]
             def update_failed?
                 @update_failed
             end
 
+            # @param [Github::PushEvent] push_event
+            # @return [void]
+            def handle_push_event(push_event, **options)
```
Comment (doudou) [inline lib/autoproj/cli/daemon.rb]: All over the pull request, it seems that `mainline` is a completely different thing that `not mainline`. It might be better to handle it this way (having two separate set of methods and hooks) ? See for instance `handle_push_event` below.

File: lib/autoproj/cli/daemon.rb [hunk 2]
```diff
@@ -37,76 +49,189 @@ def resolve_packages
...
+                if options[:mainline]
```
Comment (doudou) [inline lib/autoproj/cli/daemon.rb]: Use keyword arguments instead of option hash in this method. The option hashes are useful mostly for forwarding, or if the list of options is very dynamic.

File: lib/autoproj/cli/daemon.rb [hunk 3]
```diff
@@ -37,76 +49,189 @@ def resolve_packages
...
+                end
+            end
+
+            # @param [Github::PullRequestEvent] pull_request_event
+            # @return [void]
+            def handle_pull_request_event(pull_request_event)
+                return if update_failed?
+
+                pr = pull_request_event.pull_request
+                return if pr.base_owner == buildconf_package.owner &&
```
Comment (doudou) [inline lib/autoproj/cli/daemon.rb]: This is the second time I see this test ... maybe giving it a name ? e.g. `buildconf_pull_request?`

File: lib/autoproj/cli/daemon.rb [hunk 4]
```diff
@@ -37,76 +49,189 @@ def resolve_packages
...
+                        local_dir: package_set ? pkg[:raw_local_dir] : pkg[:srcdir]
+                    )
+                end.compact
+            end
+
+            # @return [PackageRepository]
+            def buildconf_package
+                return @buildconf_package if @buildconf_package
+
+                ws.load_config
```
Comment (doudou) [inline lib/autoproj/cli/daemon.rb]: Is there really an advantage to lazy-load the workspace ? I am really wondering ... If not, I would think that it would be best to load it early in `#initialize` and be done with it.

**Comment by g-arjones** [comment]
The difference in efficiency is about how often an event happens vs how often a pull request/mainline branch changes.. We will be polling fewer targets but events change much more often so I can't say for sure that we are really using less api calls by polling events.. Pretty much every interaction within the organization triggers some kind of event

File: lib/autoproj/daemon/buildbot.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            def validate_options(options = {})
+                options[:branch] ||= 'master'
+                options
+            end
+
+            # @param [Hash] options
+            # @return [Hash]
+            def body(**options)
+                body = BODY
+                body[:params].merge!(validate_options(options))
```
Comment (doudou) [inline lib/autoproj/daemon/buildbot.rb]: You are modifying BODY in place.

~~~
body = BODY.merge(
    params: BODY[:params].merge(branch: branch).merge(**options)
)
~~~

File: lib/autoproj/daemon/buildbot.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            # @param [Hash] options
+            # @return [Hash]
+            def validate_options(options = {})
+                options[:branch] ||= 'master'
+                options
+            end
+
+            # @param [Hash] options
+            # @return [Hash]
+            def body(**options)
```
Comment (doudou) [inline lib/autoproj/daemon/buildbot.rb:33]: The `validate_options` thing is an artifact of old days ... Nowadays you would do

~~~
branch: :master, **option
~~~

File: lib/autoproj/daemon/github_watcher.rb [hunk 7]
```diff
@@ -1,189 +1,224 @@
...
+            # @return [String] An array with all users that we will be polling
+            def owners
+                @packages.map(&:owner).uniq
             end
 
-            Repository = Struct.new :owner, :name, :branch, :timestamp
+            # @param [Array<Github::PushEvent, Github::PullRequestEvent>] events
+            # @return [(String, String)]
+            def event_owner_and_name(event)
+                if event.kind_of? Github::PushEvent
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: dispatching against class like this is usually done with `case ... when`

~~~
case event
when Github::PushEvent
when Github::PullRequestEvent
else
end
~~~

File: lib/autoproj/daemon/github_watcher.rb [hunk 8]
```diff
@@ -1,189 +1,224 @@
...
+                    owner = event.owner
+                    name = event.name
+                elsif event.kind_of? Github::PullRequestEvent
+                    owner = event.pull_request.base_owner
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: I would directly return instead of assigning to finally return.

File: lib/autoproj/daemon/github_watcher.rb [hunk 9]
```diff
@@ -1,189 +1,224 @@
...
-            # @param [String] head The head definition in the format user:branch
-            # @return [Boolean] Whether there's a PR open or not
-            def pullrequest_open?(repo, head)
-                @client.pull_requests("#{repo.owner}/#{repo.name}",
-                                      base: repo.branch).any? do |pr|
-                    pr['head']['label'] == head
+            # @param [Array<Github::PushEvent, Github::PullRequestEvent>] events
+            # @return [(Array<Github::PushEvent>, Array<Github::PullRequestEvent>)]
+            def partition_events_by_type(events)
+                push_events = events.select { |event| event.kind_of? Github::PushEvent }
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: There is `Enumerable#partition`

File: lib/autoproj/daemon/github_watcher.rb [hunk 10]
```diff
@@ -1,189 +1,224 @@
...
                 end
+                [push_events, pull_request_events]
             end
 
-            REFS_HEADS_RANGE = ('refs/heads/'.length..-1).freeze
+            # @param [Array<Github::PushEvent, Github::PullRequestEvent>] events
+            # @return [Hash]
+            def partition_events_by_repo_name(events)
+                partitioned = {}
+                events.each do |event|
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: `Enumerable#group_by`

File: lib/autoproj/daemon/github_watcher.rb [hunk 11]
```diff
@@ -1,189 +1,224 @@
...
-                        end
-                    else
-                        @push_hooks.each do |hook|
-                            hook.call("#{repo.owner}/#{repo.name}",
-                                      branch: event['payload']['ref'][REFS_HEADS_RANGE])
-                        end
+            # @param [Github::PushEvent] push_event
+            # @return [void]
+            def call_push_hooks(push_event, **options)
+                @push_hooks.each { |hook| hook.call(push_event, options) }
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: Use `**options` to forward keyword arguments. This will work until now, but is deprecated (deprecation notices will appear in 2.7, 3.0 will probably remove the possibility)

File: lib/autoproj/daemon/github_watcher.rb [hunk 12]
```diff
@@ -1,189 +1,224 @@
...
+
+            # @param [Github::PullRequestEvent] pull_request_event
+            # @return [void]
+            def call_pull_request_hooks(pull_request_event)
+                @pull_request_hooks.each { |hook| hook.call(pull_request_event) }
+            end
+
+            # @param [Github::PushEvent] push_event
+            # @return [PackageRepository, nil]
+            def package_by_push_event(push_event)
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: `packages_affected_by` would clearer to me

File: lib/autoproj/daemon/github/client.rb [hunk 13]
```diff
@@ -22,6 +24,20 @@ def pull_requests(owner, name, options = {})
                     end
                 end
 
+                # @return [Array<PushEvent, PullRequestEvent>]
+                def fetch_events(owner)
```
Comment (doudou) [inline lib/autoproj/daemon/github/client.rb:56]: I have the feeling that, all over the place, there are two distinct cases for both types of events. Maybe it would be better to separate the code paths from the source, instead of keeping them together and later on separating/dispatching them ?

File: lib/autoproj/daemon/github_watcher.rb [hunk 14]
```diff
@@ -1,189 +1,224 @@
...
+            def filter_events(events)
+                events.select do |event|
+                    if event.kind_of? Github::PushEvent
+                        to_mainline?(event.owner, event.name, event.branch) ||
+                            to_pull_request?(event.owner, event.name, event.branch)
+                    elsif event.kind_of? Github::PullRequestEvent
+                        pr = event.pull_request
+                        to_mainline?(pr.base_owner, pr.base_name, pr.base_branch)
+                    else
+                        false
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: false is implied (`if .... elsif ... end` is nil if no branch is taken)

File: lib/autoproj/daemon/github_watcher.rb [hunk 15]
```diff
@@ -1,189 +1,224 @@
...
+                packages.find do |pkg|
+                    pkg.owner == push_event.owner &&
+                        pkg.name == push_event.name &&
+                        pkg.branch == push_event.branch
+                end
+            end
+
+            # @param [Github::PushEvent] push_event
+            # @return [void]
+            def dispatch_push_event(push_event)
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb:120]: Cf. comment on `handle_push_event`

File: lib/autoproj/daemon/github_watcher.rb [hunk 16]
```diff
@@ -1,189 +1,224 @@
...
+                    push_events, pull_request_events = partition_events_by_type(
+                        repo_events
+                    )
 
-                true
+                    last_push_event = push_events.max_by(&:created_at)
+                    last_pull_request_event =
+                        pull_request_events.max_by(&:created_at)
+
+                    dispatch_push_event(last_push_event) if last_push_event
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: You are filtering on date, but then you are (in `dispatch_push_event`) having other filters and two different codepaths. Doesn't that mean that `last_push_event` could be ignored while some other events would not be ?

File: lib/autoproj/daemon/github_watcher.rb [hunk 17]
```diff
@@ -1,189 +1,224 @@
...
+                    next unless last_pull_request_event
+
+                    pr = last_pull_request_event.pull_request
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb]: it would look clearer (for symmetry) that this filtering logic ends up in a separate `dispatch_pull_request_event` method

File: lib/autoproj/daemon/package_repository.rb [hunk 18]
```diff
@@ -39,6 +47,21 @@ def buildconf?
             def branch
                 vcs[:remote_branch] || vcs[:branch] || 'master'
             end
+
+            # @return [String]
+            def head_sha
+                pkg = autobuild
+                pkg.importdir ||= local_dir # TODO: What if importdir != srcdir?
```
Comment (doudou) [inline lib/autoproj/daemon/package_repository.rb]: `importdir` is a horrible hack that was introduced at a time where I didn't want to fight back. Basically `importdir` is what the importer uses, not `srcdir`.

File: lib/autoproj/daemon/package_repository.rb [hunk 19]
```diff
@@ -39,6 +47,21 @@ def buildconf?
...
+                pkg.importer.current_remote_commit(pkg, only_local: true)
+            end
+
+            # @return [Autoproj::FakePackage]
+            def autobuild
+                return Autobuild::Package[package] if Autobuild::Package[package]
```
Comment (doudou) [inline lib/autoproj/daemon/package_repository.rb]: I'd like to get rid of `Autobuild::Package[]` at some point. Could you instead store the package in `self` ?

**Review by doudou — COMMENTED** [review]
> That being said, I believe this is good enough to do some testing "in the wild" (even though I would appreciate if you could "stress-test" it before, trying to catch some more corner cases I didn't think of).

Given that "in the wild" is "in Tidewise" right now, I think it would be good to simply install the daemon on our internal server and test it against our own repository.

**Comment by g-arjones** [comment]
I've put together a small snippet to demonstrate the concept:

```ruby
require 'octokit'
require 'faraday-http-cache'

stack = Faraday::RackBuilder.new do |builder|
    builder.use Faraday::HttpCache, serializer: Marshal,
                                    shared_cache: false

    builder.use Octokit::Response::RaiseError
    builder.adapter Faraday.default_adapter
end

Octokit.middleware = stack
$client = Octokit::Client.new

def poll_events
    $client.user_events('rock-core')
end

def poll_pull_requests(times: 1)
    times.times do
        $client.pull_requests('rock-core/buildconf')
    end
end

def benchmark(description)
    puts description
    start_time = Time.now

    yield

    puts "  Time cost: #{Time.now - start_time} secs"
    puts
end

# Do once, to build the cache
poll_events
poll_pull_requests

# Now benchmark
previous_limit = $client.rate_limit!.remaining
benchmark("Event polling:") { poll_events }
benchmark("Pull Request polling:") { poll_pull_requests(times: 50) }
puts "Total rate limit cost: #{previous_limit - $client.rate_limit!.remaining}"
```

Results:

```console
➜  ~ ruby cache_test.rb
Event polling:
  Time cost: 0.002573742 secs

Pull Request polling:
  Time cost: 0.032618024 secs

Total rate limit cost: 0
```

If we consider the difference in time cost to be negligible (I am comparing one call to fifty calls here), the real questions are:

1. How often will we have new events vs pull request changes?
2. Is the efficiency gain worth the complexity?

I don't have an answer to those questions, which makes me second guess myself even more...

**Comment by doudou** [comment]
Just to be sure, my understanding is that the `pull_request` call is also cached (something I didn't consider) .... would just need to be sure it's done properly using ETag (i.e. hitting the network at each call). In which case, I have to admit my argument doesn't really hold.

I have to admit I am surprised by this number. The "Event polling" means 2.5ms to establish a HTTPS connection with authentication ... doesn't sound right.

File: lib/autoproj/daemon/package_repository.rb (hunk 18 above)
Comment (g-arjones) [reply lib/autoproj/daemon/package_repository.rb]: That's the point.. The installation-manifest stores srcdir.. Since we can't load the workspace (we can't assume it's not broken) how do I guarantee that I am creating the importer with the same path it actually used to checkout the package?

File: lib/autoproj/daemon/package_repository.rb (hunk 19 above)
Comment (g-arjones) [reply lib/autoproj/daemon/package_repository.rb]: I could but that's not the point here.. I have to use an existing instance (if an update was made) otherwise we get an error because packages are global. What I can do is use the `find_autobuild_package` rather than `Autobuild::Package`, would that suffice?

**Comment by g-arjones** [comment]
> my understanding is that the pull_request call is also cached

Most github API calls are, that includes pull requests and branches (I've tested).

> would just need to be sure it's done properly using ETag (i.e. hitting the network at each call)

We have to look into `faraday-http-cache` docs to be sure (which we should, regardless)

File: lib/autoproj/daemon/package_repository.rb (hunk 19 above)
Comment (doudou) [reply lib/autoproj/daemon/package_repository.rb]: > would that suffice?

Yep

File: lib/autoproj/daemon/package_repository.rb (hunk 18 above)
Comment (doudou) [reply lib/autoproj/daemon/package_repository.rb]: Then add importdir to the installation manifest.

Comment (g-arjones) [reply lib/autoproj/daemon/package_repository.rb]: Sounds good.. Could you please create an issue in autoproj and assign me? Thanks

**Comment by g-arjones** [comment]
> Given that "in the wild" is "in Tidewise" right now, I think it would be good to simply install the daemon on our internal server and test it against our own repository.

Agreed.. I will update the PRs with your comments and put together a systemd service unit to push to the ansible repo. We need a build failure notification system.. Any ideas for that?

**Comment by g-arjones** [comment]
Also, it's really hard (if not impossible) to discover what triggered the build in buildbot right now. Every build looks the same both in the history and in the dashboard

**Comment by doudou** [comment]
> Every build looks the same both in the history and in the dashboard

Agreed. I think there's a way, but I don't know how.

**Comment by doudou** [comment]
> Any ideas for that?

Not really, no. Not sure what's available OOB in buildbot that would be compatible with our current workflow.

File: lib/autoproj/daemon/buildbot.rb (hunk 5 above)
Comment (g-arjones) [reply lib/autoproj/daemon/buildbot.rb]: Good catch... I wonder why this didn't raise anything even though `BODY` is frozen

File: lib/autoproj/daemon/github_watcher.rb (hunk 14 above)
Comment (g-arjones) [reply lib/autoproj/daemon/github_watcher.rb]: Yeah.. I don't like returning nil when a boolean is expected. I see this a lot in ruby code though, not sure if it makes a difference at all. I will change to keep it consistent with what you usually do

File: lib/autoproj/daemon/github_watcher.rb (hunk 16 above)
Comment (g-arjones) [reply lib/autoproj/daemon/github_watcher.rb]: Damn... The rationale was that I don't care if a PR was opened and closed several times, I'm only interested in the last event (to find out whether the PR is currently open or closed). Same goes for push (I'm only interested in the last one). The problem is that I am doing this for the entire repo when I should be doing per PR (or per branch in the case of pushes). I have to partition this further to get it right :(

Do you have a better idea to implement this?

**Comment by g-arjones** [comment]
> Not really, no. Not sure what's available OOB in buildbot that would be compatible with our current workflow.

Since we are triggering on stuff we don't control, the only thing buildbot provides out of the box that would make sense for us is mail notifications.

I think it would be relatively easy to implement ourselves something that would open/close issues on the buildconf when a build fails/pass or we could have a single issue where buildbot would report build failures through comments...

File: lib/autoproj/daemon/buildbot.rb (hunk 5 above)
Comment (doudou) [reply lib/autoproj/daemon/buildbot.rb]: `freeze` is not recursive, so modifying BODY would have raised, but BODY[:params] is another object.

File: lib/autoproj/daemon/github_watcher.rb (hunk 16 above)
Comment (doudou) [reply lib/autoproj/daemon/github_watcher.rb]: I think you need to adopt more of a data processing pipeline structure. That is, having the processing done as passing data structure(s) from functions to functions until you have the data you need to call the hooks.

One method that basically does

~~~
[get events] -> [filter/process/transform all events] -> ["collapse" redundant events] -> call hooks
~~~

where the get/filter/transform/collapse/call are processing methods implemented here and there. Most of the classes would become "passive", that is not call anything that would end up having the system actually act.

That is, in the end, making sure that the "call hooks" is done at the first level of the stack, not the last. Basically the style I mentioned in one of the first pull requests for autoproj-daemon if you remember.

**Comment by doudou** [comment]
Would it make sense to create pull requests (with a ci-do-not-merge flag) and then use http://docs.buildbot.net/current/manual/configuration/reporters.html#githubcommentpush and http://docs.buildbot.net/current/manual/configuration/reporters.html#githubstatuspush ?

**Comment by g-arjones** [comment]
I think that would be annoying.. We would be getting two notifications (one for he PR and another one for the comment). And we would get notifications for successful builds as well (not sure if we want that for packages we don't control). My concern is the we may end up just ignoring everything that comes from buildbot

File: lib/autoproj/daemon/github_watcher.rb (hunk 16 above)
Comment (g-arjones) [reply lib/autoproj/daemon/github_watcher.rb]: @doudou Is this what you mean: https://github.com/rock-core/autoproj-daemon/pull/12/commits/4b1ce9205b79f412b9c11a0f66e7bd95a73a92ab ?

**Comment by doudou** [comment]
I understand the concern ... but I don't see anything anywhere that's getting me to the conclusion that we would get all these notifications ... Could you expand a bit ?

**Comment by g-arjones** [comment]
My understanding of what you suggested is that buildbot would:

1. Open a PR (and tag it) when the daemon creates a branch in the buildconf (that is, when a PR to a mainline package/branch is opened)
2. Comment in the pull request when (at least) the build fails

Both (1) and (2) makes github send us an e-mail notification..

**Comment by doudou** [comment]
methinks the daemon would create the pull request, but anyways

We would get two notifications, but they are different notifications:
- the PR notification says "there's a new build detected", which would double with the actual pull request (from the actual package), but I don't think it doubles with the comment if the build fails.
- the build failed ... it's the whole point, isn't it ?

The one notification that's harder to not have is the pull-request-creation request. But then, we would have an idea using issues as you originally suggested. I personally would prefer to not make email mandatory (I don't like notification emails)

We picked buildbot originally because it's relatively easy to "tune", and that's one case where it is. The `filterBuilds` method on the reporter is called to decide whether a notification should or not be sent (https://github.com/buildbot/buildbot/blob/master/master/buildbot/reporters/http.py#L71)
 
Given the general effort spent, I'd say we go for it and tune later ? The PR creation notification *will* be useful at first to track what the daemon does.


Maybe the draft pull requests would help ?

**Comment by g-arjones** [comment]
> methinks the daemon would create the pull request

Agreed.

The draft pull request is an interesting idea.. I have never used them but maybe that would help indeed. OK, let's go for PR + comment then.

File: lib/autoproj/daemon/github/client.rb (hunk 13 above)
Comment (g-arjones) [reply lib/autoproj/daemon/github/client.rb:56]: Not sure.. That would make sense now since we are only interested in PushEvent and PullRequestEvent but we will soon be fetching comments (for the hashtag thing) and maybe more later on.. We could group them here but I am not sure if that makes sense in the client API level..

File: test/autoproj/cli/test_daemon.rb [hunk 20]
```diff
@@ -33,69 +37,442 @@ def define_package(name, vcs)
...
-                    flexmock(Process)
-                        .should_receive(:exec)
-                        .with(Gem.ruby, $PROGRAM_NAME, 'daemon',
-                              'start', '--update').once
 
                     watcher.should_receive(:add_push_hook)
-                           .and_yield('rock-core', 'autoproj')
+                           .and_yield('push_event', options: 'options')
+                    watcher.should_receive(:add_pull_request_hook)
+                           .and_yield('pull_request_event')
```
Comment (doudou) [inline test/autoproj/cli/test_daemon.rb:473]: This use the incidental fact that `#setup_hook` passes a block to `add_push_hook` and that therefore making `add_push_hook` to yield will cause `handle_push_event` to be called. Which is not its role.

IMO, what should be done is explicitly call `call_push_hooks` from the test itself.

File: test/autoproj/daemon/github/test_client.rb [hunk 21]
```diff
@@ -33,6 +34,23 @@ module Github
...
+                        { 'type' => 'CreateEvent' }
+                    ]
+
+                    flexmock(Octokit::Client).new_instances.should_receive(:user_events)
+                                             .with('rock-core').and_return(events)
+
+                    @client = Client.new(auto_paginate: false)
+                    fetched_events = client.fetch_events('rock-core')
+                    assert_equal 2, fetched_events.size
+                    assert fetched_events.first.kind_of? PushEvent
```
Comment (doudou) [inline test/autoproj/daemon/github/test_client.rb]: `assert_kind_of`

File: test/autoproj/daemon/github/test_pull_request_event.rb [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+                before do
+                    @model = JSON.parse(
+                        File.read(File.expand_path('pull_request_event.json', __dir__))
+                    )
+
+                    @pull_request_event = PullRequestEvent.new(@model)
+                end
+
+                it 'returns the state' do
+                    assert true, pull_request_event.pull_request.open?.class
```
Comment (doudou) [inline test/autoproj/daemon/github/test_pull_request_event.rb]: This is `assert true`, a.k.a. always passes

File: lib/autoproj/daemon/github_watcher.rb [hunk 23]
```diff
@@ -1,189 +1,213 @@
...
+            # @return [void]
+            def handle_owner_events(events)
+                events = filter_events(events)
+                push_events, pull_request_events =
+                    events.partition { |event| event.kind_of? Github::PushEvent }
+
+                handle_push_events(push_events) unless push_events.empty?
+                return if pull_request_events.empty?
+
+                handle_pull_request_events(pull_request_events)
```
Comment (doudou) [inline lib/autoproj/daemon/github_watcher.rb:150]: There's no real performance need to have shortcuts here, and the `handle_*` methods should be able to cleanly handle empty arguments (they do). I don't think there should be a special case for an empty array.

**Comment by g-arjones** [comment]
There is a problem with using `Autoproj::Manifest#find_autobuild_package`:

```ruby
ws = Autoproj.workspace
ws.manifest.find_autobuild_package('drivers/iodrivers_base') # => nil

Autoproj::CLI::Main.start(%w[envsh])
Autoproj.workspace.manifest.find_autobuild_package(
    'drivers/iodrivers_base'
) # => #<Autobuild::CMake name=drivers/iodrivers_base>

ws.manifest.find_autobuild_package('drivers/iodrivers_base') # => nil
```

The problem stems from the fact that `Autoproj::CLI::Base` [uses `Autoproj::Workspace.default` to get workspace instances](https://github.com/rock-core/autoproj/blob/83703e6098442320e370c27d15ffe7c28993d8dc/lib/autoproj/cli/base.rb#L16), which [always creates a new one](https://github.com/rock-core/autoproj/blob/83703e6098442320e370c27d15ffe7c28993d8dc/lib/autoproj/workspace.rb#L107) rather than return the one that was set to default with `Workspace#set_as_main_workspace`. The "fix" is rather obvious (either use `Autoproj.workspace` to get instances in `Autoproj::CLI::Base` or make `Autoproj::Workspace.default` always return the same) but I am not sure about how far reaching are the consequences of this. 

Or, I could just always use `Autoproj.workspace` in the daemon and ignore the underlying issue but I understand you are trying to steer away from that in the codebase. Thoughs?

**Comment by g-arjones** [comment]
> There is a problem with using `Autoproj::Manifest#find_autobuild_package`:
>
> ```ruby
> ws = Autoproj.workspace
> ws.manifest.find_autobuild_package('drivers/iodrivers_base') # => nil
>
> Autoproj::CLI::Main.start(%w[envsh])
> Autoproj.workspace.manifest.find_autobuild_package(
>     'drivers/iodrivers_base'
> ) # => #<Autobuild::CMake name=drivers/iodrivers_base>
>
> ws.manifest.find_autobuild_package('drivers/iodrivers_base') # => nil
> ```
>
> The problem stems from the fact that `Autoproj::CLI::Base` [uses `Autoproj::Workspace.default` to get workspace instances](https://github.com/rock-core/autoproj/blob/83703e6098442320e370c27d15ffe7c28993d8dc/lib/autoproj/cli/base.rb#L16), which [always creates a new one](https://github.com/rock-core/autoproj/blob/83703e6098442320e370c27d15ffe7c28993d8dc/lib/autoproj/workspace.rb#L107) rather than return the one that was set to default with `Workspace#set_as_main_workspace`. The "fix" is rather obvious (either use `Autoproj.workspace` to get instances in `Autoproj::CLI::Base` or make `Autoproj::Workspace.default` always return the same) but I am not sure about how far reaching are the consequences of this. 
>
> Or, I could just always use `Autoproj.workspace` in the daemon and ignore the underlying issue but I understand you are trying to steer away from that in the codebase. Thoughs?

@doudou Not sure if you missed this one...

**Comment by doudou** [comment]
> @doudou Not sure if you missed this one...

I did. Can you point to where in the code this happens exactly, I grepped for CLI::Main.start but could not find anything.

**Comment by g-arjones** [comment]
> Can you point to where in the code this happens exactly

https://github.com/rock-core/autoproj-daemon/blob/af36c12da2ef3c84ed2161ca9b58fa78386db22f/lib/autoproj/cli/daemon.rb#L286

File: test/autoproj/cli/test_daemon.rb (hunk 20 above)
Comment (g-arjones) [reply test/autoproj/cli/test_daemon.rb:473]: This is not incidental, this is actually the very thing I want to test here. The only purpose of `setup_hooks` is to call `add_push_hook` and `add_pull_request_hook` passing a block that forwards the arguments to the proper handlers.. This *is* its role. I could, rather than yielding, store the block in the test case and "force" call it to check that the expectations are met but I don't see any difference in that.

`call_push_hooks` is a method of `GithubWatcher` which is not the test subject, calling that should be done in `GithubWatcher`'s tests (and it is).

Comment (doudou) [reply test/autoproj/cli/test_daemon.rb:473]: Well. `add_push_hook` and `add_pull_request_hook` _in real life_ do not yield. Making them yield here is at the very least very confusing (this discussion being a consequence of that). You're basically mis-using `and_yield` to verify what block is being passed as argument to `add_push_hook`.

My understanding is that you are trying to be very unit-testy and not touch anything but "your subject". I don't subscribe to this strategy, but let's keep on with it regardless.

There is no easy way to test the block passed to a method. If you do control the block, you can. But you're not controlling it in this case ... so `and_yield` is indeed a good solution - just put a lot of commenting to explain why `and_yield` is being used here on a method that never yields.

Now, I believe that doing it through `call_push_event` would be even better, as it tests that `setup_hooks` does _whatever is necessary_ to be called when the watcher has an event. It is  less strictly "unit-testy", but IMO performs a better test as it assumes less about the underlying APIs and more about the actual functionality / behavior that is required from it.

**Comment by doudou** [comment]
About the `Main.start` thingy: you could use CLI::Update directly, which takes the workspace as constructor.

File: test/autoproj/cli/test_daemon.rb (hunk 20 above)
Comment (g-arjones) [reply test/autoproj/cli/test_daemon.rb:473]: > you are trying to be very unit-testy

We've been in this position before and we really do have different opinions on this. There's definitely an argument to both sides. I honestly believe that by testing a lot of stuff in a test case  hurts maintainability of the tests. It also usually introduces dependencies on implementation details (specially if the resources requires mocks).

So, if you don't mind I will just add the comment you suggested...

**Comment by g-arjones** [comment]
> you could use CLI::Update directly

Humm.. Good idea, I will give that a try

File: test/autoproj/cli/test_daemon.rb (hunk 20 above)
Comment (doudou) [reply test/autoproj/cli/test_daemon.rb:473]: > So, if you don't mind I will just add the comment you suggested...

I don't, this is why I suggested them. And I agree that we have to continue disagreeing on this.

Having a lot of mocks is to me a test smell and a test maintainability problem - something I'm not the only vocal about in the industry, and in this case they could have been very simply reduced. We've already had that discussion because you were producing tests with *literally* pages of mocks. This is *not* helping maintainability.

**Review by doudou — APPROVED** [review]
